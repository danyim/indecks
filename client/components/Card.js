import React from 'react';
import { Link } from 'react-router';
import CSSTransitionGroup from 'react-addons-css-transition-group';

const Card = React.createClass({
  render() {
    const { card } = this.props;

    return (
      <figure className="grid-figure">
        <div className="card-title">
          <h1>{card.title}</h1>
        </div>

        <figcaption>
          <p>{card.answer}</p>
        </figcaption>
      </figure>
    );
  }
})

export default Card;
