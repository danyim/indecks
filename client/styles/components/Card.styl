.card-title
  font-size 2.4rem
  cursor pointer
