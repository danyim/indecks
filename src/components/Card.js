import React from 'react';
import { Link } from 'react-router';
import CSSTransitionGroup from 'react-addons-css-transition-group';
import Markdown from './Markdown';
import styles from '../styles/components/Card';

class Card extends React.Component {
  render() {
    const { card, flipped } = this.props;

    let emptyAnswer;
    if(card.answer === '') {
      emptyAnswer =
        <div className={`${styles['center']}`}>
          <p className={`${styles['grey']}`}>No answer provided</p>
        </div>;
    }

    return (
      <figure className={`grid-figure ${styles['grid-figure']}`}>
        {(() => {
          if(flipped === false) {
            return (
              <div onClick={this.props.handleOnClick || null}>
                <Markdown className={`${styles['card-title']}`} text={card.title} removeParaTags={true} />
              </div>
            );
          }
          else {
            return (
              <figcaption>
                <Markdown text={card.answer} />
                {emptyAnswer}
              </figcaption>
            );
          }
        })()}
      </figure>
    );
  }
}

Card.defaultProps = {
  flipped: false
};
Card.propTypes = {
  card: React.PropTypes.object.isRequired,
  flipped: React.PropTypes.bool
};

export default Card;
