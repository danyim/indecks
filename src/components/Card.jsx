import React from 'react'
import PropTypes from 'prop-types'
import { CardShape } from './__commonShapes'
import Markdown from './Markdown'
import styles from '../styles/components/Card.styl'

class Card extends React.Component {
  static propTypes = {
    card: CardShape.isRequired,
    className: PropTypes.string,
    flipped: PropTypes.bool,
    handleOnClick: PropTypes.func,
    trimOverflow: PropTypes.bool,
    trimOverflowLength: PropTypes.number,
    children: PropTypes.node,
  }

  static defaultProps = {
    className: '',
    flipped: false,
    handleOnClick: () => {},
    trimOverflow: false,
    trimOverflowLength: 125,
    children: null,
  }

  static renderEmpty() {
    return (
      <div className={`${styles.center}`}>
        <p className={`${styles.grey}`}>This card does not have an answer yet</p>
      </div>
    )
  }

  static renderHasAnswerIcon(card) {
    if (!(card && card.answer && card.answer.trim() !== '')) {
      return (
        <div className={`${styles['no-answer']}`} title="This card has no answer">
          <div>No answer</div>
        </div>
      )
    }
    return null
  }

  constructor(props) {
    super(props)

    this.trimOverflowTitle.bind(this, this.trimOverflowTitle)
  }

  trimOverflowTitle(text) {
    const textWithoutImg = text.replace(/!\[.*\]\(.*\)/g, '')
    return this.props.trimOverflow && textWithoutImg.length > this.props.trimOverflowLength
      ? `${textWithoutImg.substr(0, this.props.trimOverflowLength)}...`
      : text
  }

  renderMarkdown() {
    const { card, flipped } = this.props
    if (flipped === false) {
      return (
        <Markdown
          className={`card-title ${styles['card-title']}`}
          text={this.trimOverflowTitle(card.title)}
        />
      )
    }
    return (
      <figcaption>
        {card.answer !== null && card.answer !== '' ? (
          <Markdown text={card.answer} />
        ) : (
          Card.renderEmpty()
        )}
      </figcaption>
    )
  }

  render() {
    const { card, className, handleOnClick, children } = this.props

    return (
      <figure
        className={`grid-figure ${styles['grid-figure']} ${className}`}
        onClick={() => handleOnClick()}
        role="presentation"
      >
        <div className={`card-contents ${styles['card-contents']}`}>{this.renderMarkdown()}</div>
        {Card.renderHasAnswerIcon(card)}
        {children}
      </figure>
    )
  }
}

export default Card
