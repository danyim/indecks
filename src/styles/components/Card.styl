@import '../_variables'

figure.grid-figure
  min-height 250px
  flex-direction column
  figcaption
    padding 0rem 2rem 2rem 2rem
    li
      font-size 1.5rem
      line-height 1.5

.card-title
  font-size 2.5rem
  padding 2rem 2rem 1rem 2rem
  cursor pointer

.center
  justify-content: center
  align-content: center
  p
    color lightgray
    font-weight bold
    text-align center
