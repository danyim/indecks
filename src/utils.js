export const createReducer = (initialState, handlers) => {
  return function reducer(state = initialState, action = {}) {
    if (handlers.hasOwnProperty(action.type)) {
      return handlers[action.type](state, action);
    }
    return state;
  };
};

/**
 * Given an array of consecutive numbers, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a Array of numbers
 * @return {Number}   Missing value
 */
export const findMissing = (a) => {
  let expectedNextValue;
  for(let k of a)
  {
    expectedNextValue = a[k] + 1;
    if(k <= a.length - 1 && expectedNextValue !== a[k + 1]) {
      return expectedNextValue;
    }
  }
  return null;
};

/**
 * Given an array with potentially missing values at parameter key, find the
 * next best 0-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) => {
  return findMissing(a.map(x => x[key])) || Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1);
};

