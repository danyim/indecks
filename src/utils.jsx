import React from 'react'
import { Route } from 'react-router-dom'
import _ from 'lodash'

export const createReducer = (initialState, handlers) =>
  function reducer(state = initialState, action = {}) {
    if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
      return handlers[action.type](state, action)
    }
    return state
  }

/**
 * Given an array of consecutive numbers from 1 to n, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a   Array of numbers
 * @param  {Number} min Minimum value (0 or 1)
 * @return {Number}     Missing value
 */
export const findMissing = (a, min = 1) => {
  let expected = min
  for (let k = 0; k < a.length; k++) {
    if (a[k] !== expected) {
      return expected
    }
    expected++
  }
  return null
}

/**
 * Given an array of objects, check the value of the key and return the
 * best 1-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) =>
  findMissing(a.map(x => x[key])) ||
  Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1)

/**
 * Helper function to generate routes
 * @param  {Object} route) Route object that takes the form
 *                         {
 *                           exact: boolean,
 *                           path: string,
 *                           component: Component,
 *                           routes: Routes
 *                         }
 * @return {Component}        Returns a <Route> component to plug into a <Switch>
 */
export const routeWithSubRoutes = route =>
  <Route
    key={_.uniqueId()}
    exact={route.exact || false}
    path={route.path}
    render={props =>
      // Pass the sub-routes down to keep nesting
      // <route.component {...props}>
      //   { route.children ? route.children.map(child => routeWithSubRoutes(child)) : null }
      // </route.component>
      <route.component {...props} routes={route.routes} />}
  />
