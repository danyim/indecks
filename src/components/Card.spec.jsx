import toJSON from 'enzyme-to-json'
import { shallow, mount, setup, setupFull } from '../test/utils' // eslint-disable-line no-unused-vars
import Card from './Card'

describe('Card', () => {
  let props

  beforeEach(() => {
    props = {
      card: {
        title: 'Title',
        answer: 'Answer',
        index: 0,
      },
      className: '',
      flipped: false,
      handleOnClick: jest.fn(),
      trimOverflow: false,
      trimOverflowLength: 125,
      children: null,
    }
  })
  it('should render self and subcomponents', () => {
    const wrapper = mount(Card, props)
    expect(toJSON(wrapper)).toMatchSnapshot()
  })

  it('should call the handleOnClick once when card is clicked', () => {
    const handler = jest.fn()
    const { wrapper } = setup(Card, {
      ...props,
      handleOnClick: handler,
    })

    wrapper.find('figure').simulate('click')
    expect(handler.mock.calls.length).toBe(1)
  })

  it('should render the title if flipped is FALSE', () => {
    const { wrapper } = setup(Card, {
      ...props,
      flipped: false,
    })

    expect(wrapper.find('Markdown.card-title').exists()).toEqual(true)
  })

  it('should render the answer if flipped is TRUE', () => {
    const { wrapper } = setup(Card, {
      ...props,
      flipped: true,
    })

    expect(wrapper.find('figcaption Markdown').exists()).toEqual(true)
  })

  it('should render the no answer indicator on the title if a null or blank answer is provided', () => {
    const { wrapper } = setup(Card, {
      ...props,
      card: {
        ...props.card,
        answer: '',
      },
    })

    expect(wrapper.find('div.no-answer > div').text()).toBe('No answer')
  })

  it('should render an empty answer notice if the card is flipped', () => {
    const { wrapper } = setup(Card, {
      ...props,
      card: {
        ...props.card,
        answer: '',
      },
      flipped: true,
    })

    expect(wrapper.find('figcaption Markdown').exists()).toEqual(false)
    expect(wrapper.find('div.center p.grey').text()).toEqual(
      'This card does not have an answer yet'
    )
  })

  it('should trim the title if it exceeds n chars', () => {
    const title = 'Testing long title'
    const trimOverflowLength = 5

    const { wrapper } = setup(Card, {
      ...props,
      card: {
        ...props.card,
        title,
      },
      flipped: false,
      trimOverflow: true,
      trimOverflowLength,
    })

    expect(
      wrapper.find(`Markdown[text="${title.substr(0, trimOverflowLength)}..."]`).exists()
    ).toEqual(true)
  })

  it('should trim the title not including markdown image tags', () => {
    const extra = '![Image title](http://image.url)'
    const title = 'Testing long title with an image'
    const trimOverflowLength = 5

    const { wrapper } = setup(Card, {
      ...props,
      card: {
        ...props.card,
        title: extra + title,
      },
      flipped: false,
      trimOverflow: true,
      trimOverflowLength,
    })

    expect(
      wrapper.find(`Markdown[text="${title.substr(0, trimOverflowLength)}..."]`).exists()
    ).toEqual(true)
  })
})
