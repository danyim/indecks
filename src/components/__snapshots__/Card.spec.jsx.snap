// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render self and subcomponents 1`] = `
<Card
  card={
    Object {
      "answer": "Answer",
      "index": 0,
      "title": "Title",
    }
  }
  className=""
  flipped={false}
  handleOnClick={[MockFunction]}
  trimOverflow={false}
  trimOverflowLength={125}
>
  <figure
    className="grid-figure grid-figure "
    onClick={[Function]}
    role="presentation"
  >
    <div
      className="card-contents card-contents"
    >
      <Markdown
        className="card-title card-title"
        handleOnClick={null}
        text="Title"
      >
        <div
          className="card-title card-title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Title</p>
",
            }
          }
          onClick={null}
        />
      </Markdown>
    </div>
  </figure>
</Card>
`
